incubator/etcd/templates/configmap.yaml

```diff
@@ -0,0 +1,176 @@
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+data:
+  init.sh: |
+    #!/bin/sh
+    set -ex
+
+    HOSTNAME=$(hostname)
+    IP=$(ip r get 1 | awk '{print $NF;exit}')
+    PROTO=http{{- if .Values.etcd.peerTLS }}s{{- end }}
+
+    # store member id into PVC for later member replacement
+    collect_member() {
+      while ! etcdctl member list &>/dev/null; do sleep 1; done
+      etcdctl member list | grep http://${HOSTNAME}.${SET_NAME}.${POD_NAMESPACE}:2379 | cut -d':' -f1 | cut -d'[' -f1 > /etcd/member_id
+      exit 0
+    }
+
+    # endpoint of the first etcd member
+    ep() {
+      echo "http://${SET_NAME}-0.${SET_NAME}.${POD_NAMESPACE}:2379"
+    }
+
+    # wait for the given hostname to become available to the network
+    await_host() {
+      echo "Waiting for [${1}] to come up "
+      while true; do
+        echo -n '.'
+        ping -W 1 -c 1 ${1} > /dev/null && break
+        sleep 1s
+      done
+      echo -n " done"
+    }
+
+    # add a new member to the etcd cluster
+    add_member() {
+      EPS=$(ep)
+      MEMBER_LIST=$(etcdctl --endpoints ${EPS} member list)
+
+      if echo "${MEMBER_LIST}" | grep ${HOSTNAME}; then
+        echo "Removing old member [${HOSTNAME}] from etcd cluster"
+        MEMBER_ID=$(echo "$MEMBER_LIST" | grep ${HOSTNAME} | cut -d':' -f1 | cut -d'[' -f1)
+        etcdctl --endpoints ${EPS} member remove ${MEMBER_ID}
+      fi
+
+      if [ -e $ETCD_DATA_DIR ]; then
+        echo "Removing old data dir [${ETCD_DATA_DIR}] "
+        rm -Rf $ETCD_DATA_DIR
+      fi
+
+      echo "Adding [${HOSTNAME}] as a new member"
+      STR_ENV="ETCD_"
+      ETCD_PREFIX=$(etcdctl --endpoints ${EPS} member add ${HOSTNAME} ${PROTO}://${HOSTNAME}.${SET_NAME}.${POD_NAMESPACE}:2380 | \
+      {
+        while read i
+          do
+            if test "${i#*$STR_ENV}" != "$i"; then
+              ETCD_PREFIX="$ETCD_PREFIX $i"
+            fi
+          done
+        echo "export $ETCD_PREFIX" | sed 's/"//g'
+      })
+      eval \${ETCD_PREFIX}
+
+      collect_member &
+      echo "Start etcd daemon"
+      exec etcd \
+        --listen-peer-urls ${PROTO}://${IP}:2380 \
+        --initial-advertise-peer-urls ${PROTO}://${HOSTNAME}.${SET_NAME}.${POD_NAMESPACE}:2380 \
+        --listen-client-urls http://${IP}:2379,http://127.0.0.1:2379 \
+        {{- if .Values.etcd.peerTLS }}
+        --peer-auto-tls \
+        {{- end }}
+        --advertise-client-urls http://${HOSTNAME}.${SET_NAME}.${POD_NAMESPACE}:2379
+    }
+
+    # <release-name>-etcd-<set-id>
+    SET_ID=${HOSTNAME##*-}
+
+    # are we rejoining after a failure?
+    if [ -e /etcd/member_id ]; then
+        echo "Found old member ID, attempting to rejoin the etcd cluster"
+
+        MEMBER_ID=$(cat /etcd/member_id)
+
+        # if we are the first member we can't resolve over the default endpoint
+        # attempt to resolve over the defined service instead
+        if [ "${SET_ID}" -eq 0 ]; then
```

Comment:
If you'd do a `kubectl delete pod etcd-0` and k8s spawned a new Pod it would create a new cluster. This `if/else` block fixes it by falling back to the defined service if the `SET_ID is 0`.